### PR TITLE
[SPARC][IAS] Emit the correct 32-bit ELF machine type

### PR DESCRIPTION
**Pull request by koachan**
Emit the correct machine type when writing out 32-bit ELF objects. This patch is modeled on GCC's behavior:
- `-m32` emits an object of type EM_SPARC;
- `-m32 -mcpu=v9` emits EM_SPARC32PLUS (however, see below); and
- `-m64` emits EM_SPARCV9.

Note:
GNU as doesn't actually support user control of emitted machine type. It will always autodetect the type based on the instruction mix:
- If there's a V9 instruction inside, then use EM_SPARC32PLUS; and
- Emit EM_SPARC otherwise.

GCC simply happens to try to emit V9 instructions whenever possible with `-m32 -mcpu=v9`, so there's a high chance that the resulting object file will be of a EM_SPARC32PLUS type, however it is not a guaranteed behavior.

For this patch, I opt for deterministic behavior for simplicity.

### CODE REVIEW

File: llvm/lib/Target/Sparc/MCTargetDesc/SparcAsmBackend.cpp [hunk 1]
```diff
@@ -132,14 +132,18 @@ namespace {
...
 
   public:
-    SparcAsmBackend(const Target &T)
+    SparcAsmBackend(const Target &T, const MCSubtargetInfo &STI)
         : MCAsmBackend(StringRef(T.getName()) == "sparcel"
                            ? llvm::endianness::little
                            : llvm::endianness::big),
-          TheTarget(T), Is64Bit(StringRef(TheTarget.getName()) == "sparcv9") {}
+          TheTarget(T), TheSTI(STI),
+          Is64Bit(StringRef(TheTarget.getName()) == "sparcv9"),
```
Comment (glaubitz) [inline llvm/lib/Target/Sparc/MCTargetDesc/SparcAsmBackend.cpp:145]: Unrelated to your change, but I wonder whether checking for the target name being `sparcv9` is the proper and robust way to check whether we're on 64-bit SPARC. On Linux, the 64-bit SPARC target is called `sparc64` and not `sparcv9`.

**Comment by glaubitz** [comment]
Looks good to me. Thanks for fixing this!

**Comment by glaubitz** [comment]
> Also, any ideas on how should I go to unit test this change?

Maybe using the `file` command to test for the machine type of a generated object file?

**Comment by s-barannikov** [comment]
> Also, any ideas on how should I go to unit test this change?

Look for existing tests using `llvm-readobj -h`.

File: llvm/lib/Target/Sparc/MCTargetDesc/SparcAsmBackend.cpp [hunk 2]
```diff
@@ -132,14 +132,18 @@ namespace {
   class SparcAsmBackend : public MCAsmBackend {
   protected:
     const Target &TheTarget;
```
Comment (jrtc27) [inline llvm/lib/Target/Sparc/MCTargetDesc/SparcAsmBackend.cpp:134]: This seems unused beyond in the constructor?

File: llvm/lib/Target/Sparc/MCTargetDesc/SparcAsmBackend.cpp (hunk 1 above)
Comment (jrtc27) [reply llvm/lib/Target/Sparc/MCTargetDesc/SparcAsmBackend.cpp:145]: Now you have STI, you can just get the triple and avoid ugly string comparisons for all these (isArch64Bit and isLittleEndian)

Comment (glaubitz) [reply llvm/lib/Target/Sparc/MCTargetDesc/SparcAsmBackend.cpp:145]: Okay, thanks for the clarification. We're good then. ;-)

**Comment by glaubitz** [comment]
Unit test looks good to me as well. I would prefer squashing the two additional commits into the original commits though since the test belongs to the change.

**Comment by glaubitz** [comment]
> > Unit test looks good to me as well. I would prefer squashing the two additional commits into the original commits though since the test belongs to the change.
> 
> Yep, it'll be squashed automatically during the merge :)

OK, I wasn't aware of that.

**Review by s-barannikov — APPROVED** [review]
The changes LGTM as a step towards supporting V8+

File: llvm/lib/Target/Sparc/MCTargetDesc/SparcAsmBackend.cpp [hunk 3]
```diff
@@ -131,15 +131,16 @@ static unsigned getFixupKindNumBytes(unsigned Kind) {
...
   class SparcAsmBackend : public MCAsmBackend {
   protected:
-    const Target &TheTarget;
     bool Is64Bit;
+    bool HasV9;
 
   public:
-    SparcAsmBackend(const Target &T)
-        : MCAsmBackend(StringRef(T.getName()) == "sparcel"
+    SparcAsmBackend(const Target &T, const MCSubtargetInfo &STI)
```
Comment (jrtc27) [inline llvm/lib/Target/Sparc/MCTargetDesc/SparcAsmBackend.cpp:138]: T is unused

Comment (jrtc27) [reply llvm/lib/Target/Sparc/MCTargetDesc/SparcAsmBackend.cpp]: Yes, that's normal, you just take what you need

**Comment by s-barannikov** [comment]
> Hmm, so how do I rerun the tests?

Try force pushing or updating the branch (there is a button below CI checks).

**Review by s-barannikov — APPROVED** [review]

**Pull request merged**